File: Creational_Patterns/Factory_Pattern/java/UI_theme_system.java
import java.util.*;

interface Button{
  void render();
}

interface CheckBox{
  void select();
}

class LightButton implements Button{
  @Override
  public void render(){
    System.out.println("Rendering Light Button");
  }
}

class DarkButton implements Button{
  @Override
  public void render(){
     System.out.println("Rendering Dark Button");
  }
}

class LightCheckBox implements CheckBox{
  @Override
  public void select(){
    System.out.println("Selecting Light CheckBox");
  }
}


class DarkCheckBox implements CheckBox{
  @Override
  public void select(){
    System.out.println("Selecting Dark CheckBox");
  }
}

interface ThemeFactory{
  Button createButton();
  CheckBox createCheckBox();
}

class LightThemeFactory implements ThemeFactory{
  @Override
  public Button createButton(){
    return new LightButton();
  }

  @Override
  public CheckBox createCheckBox(){
    return new LightCheckBox();
  }
}


class DarkThemeFactory implements ThemeFactory{
  @Override
  public Button createButton(){
    return new DarkButton();
  }

  @Override
  public CheckBox createCheckBox(){
    return new DarkCheckBox();
  }
}


class ThemeService{
  HashMap<String,ThemeFactory> registry= new HashMap<>();

  ThemeFactory factory = null;

  public ThemeService(String theme){
    registry.put("light",new LightThemeFactory());
    registry.put("dark",new DarkThemeFactory());
    this.factory = registry.get(theme);
  }

  public void renderButton(){
    factory.createButton().render();
  }

  public void renderCheckBox(){
    factory.createCheckBox().select();
  }

  public void renderAll(){
    renderButton();
    renderCheckBox();
  }
}

class Main{
  public static void main(String[] args){
    ThemeService ts = new ThemeService("dark");
    ts.renderAll();
  }
}

File: Creational_Patterns/Factory_Pattern/python/UI_theme_system.py
from abc import ABC, abstractmethod

# Abstract Products
class Button(ABC):
    @abstractmethod
    def render(self): pass

class CheckBox(ABC):
    @abstractmethod
    def select(self): pass

# Concrete Products
class LightButton(Button):
    def render(self): print("Rendering Light Button")

class DarkButton(Button):
    def render(self): print("Rendering Dark Button")

class LightCheckBox(CheckBox):
    def select(self): print("Selecting Light Checkbox")

class DarkCheckBox(CheckBox):
    def select(self): print("Selecting Dark Checkbox")


# Abstract Factory
class ThemeFactory(ABC):
    @abstractmethod
    def create_button(self) -> Button: pass
    @abstractmethod
    def create_checkbox(self) -> CheckBox: pass

# Concrete Factories
class LightThemeFactory(ThemeFactory):
    def create_button(self): return LightButton()
    def create_checkbox(self): return LightCheckBox()

class DarkThemeFactory(ThemeFactory):
    def create_button(self): return DarkButton()
    def create_checkbox(self): return DarkCheckBox()


# ThemeService — fully resolves everything internally
class ThemeService:
    _registry = {
        "light": LightThemeFactory,
        "dark": DarkThemeFactory,
    }

    def __init__(self, theme: str):
        theme = theme.lower()
        if theme not in self._registry:
            raise ValueError(f"Unknown theme: '{theme}'")
        self._factory = self._registry[theme]()  # resolved once, stored

    def render_button(self):
        self._factory.create_button().render()

    def render_checkbox(self):
        self._factory.create_checkbox().select()

    def render_all(self):
        self.render_button()
        self.render_checkbox()


# Client — only knows "light" or "dark", nothing else
if __name__ == "__main__":
    ThemeService("light").render_all()
    ThemeService("dark").render_all()